File: common/cpp/public_metadata/public_metadata_test.cc
#include "privacy/net/common/cpp/public_metadata/public_metadata.h"

#include <cstdint>

#include "testing/base/public/gmock.h"
#include "testing/base/public/gunit.h"
#include "third_party/absl/time/clock.h"
#include "third_party/absl/time/time.h"

namespace privacy::ppn {
namespace {

TEST(BinaryPublicMetadataSerialize, RoundtripV1) {
  BinaryPublicMetadata metadata;
  metadata.version = 1;
  metadata.service_type = "chromeipblinding";
  metadata.country = "US";
  metadata.region = "US-CA";
  metadata.city = "SUNNYVALE";
  metadata.debug_mode = 0;
  // Round to the next 15-minute cutoff.
  uint64_t seconds = absl::ToUnixSeconds(absl::Now() + absl::Minutes(15));
  seconds -= (seconds % 900);
  metadata.expiration_epoch_seconds = seconds;
  const auto encoded = Serialize(metadata);
  ASSERT_TRUE(encoded.ok()) << encoded.status();
  const auto decoded = Deserialize(encoded.value());
  ASSERT_TRUE(decoded.ok()) << decoded.status();
  EXPECT_EQ(metadata.version, decoded.value().version);
  EXPECT_EQ(metadata.service_type, decoded.value().service_type);
  EXPECT_EQ(metadata.country, decoded.value().country);
  EXPECT_EQ(metadata.region, decoded.value().region);
  EXPECT_EQ(metadata.city, decoded.value().city);
  EXPECT_EQ(metadata.debug_mode, decoded.value().debug_mode);
  EXPECT_EQ(metadata.expiration_epoch_seconds,
            decoded.value().expiration_epoch_seconds);
}

TEST(BinaryPublicMetadataSerialize, RoundtripV2) {
  BinaryPublicMetadata metadata;
  metadata.version = 2;
  metadata.service_type = "chromeipblinding";
  metadata.country = "US";
  metadata.region = "US-CA";
  metadata.city = "SUNNYVALE";
  metadata.proxy_layer = 1;
  metadata.debug_mode = 0;
  // Round to the next 15-minute cutoff.
  uint64_t seconds = absl::ToUnixSeconds(absl::Now() + absl::Minutes(15));
  seconds -= (seconds % 900);
  metadata.expiration_epoch_seconds = seconds;
  const auto encoded = Serialize(metadata);
  ASSERT_TRUE(encoded.ok()) << encoded.status();
  const auto decoded = Deserialize(encoded.value());
  ASSERT_TRUE(decoded.ok()) << decoded.status();
  EXPECT_EQ(metadata.version, decoded.value().version);
  EXPECT_EQ(metadata.service_type, decoded.value().service_type);
  EXPECT_EQ(metadata.country, decoded.value().country);
  EXPECT_EQ(metadata.region, decoded.value().region);
  EXPECT_EQ(metadata.city, decoded.value().city);
  EXPECT_EQ(metadata.debug_mode, decoded.value().debug_mode);
  EXPECT_EQ(metadata.proxy_layer, decoded.value().proxy_layer);
  EXPECT_EQ(metadata.expiration_epoch_seconds,
            decoded.value().expiration_epoch_seconds);
}

}  // namespace
}  // namespace privacy::ppn

File: common/cpp/public_metadata/go/public_metadata_test.go
package binarymetadata

import (
	"errors"
	"testing"
	"time"

	"google3/util/task/go/status"

	tpb "google3/google/protobuf/timestamp_go_proto"
	plpb "google3/privacy/net/common/proto/proxy_layer_go_proto"
	pmpb "google3/privacy/net/common/proto/public_metadata_go_proto"
)

func TestValidateMetadataCardinality(t *testing.T) {
	bs := New(&NewBinaryFields{
		Version:     1,
		Country:     "US",
		Region:      "US-CA",
		City:        "SUNNYVALE",
		ServiceType: "chromeipblinding",
		Expiration:  tpb.New(time.Now().Add(time.Minute * 30).Round(time.Minute * 15)),
		DebugMode:   pmpb.PublicMetadata_DEBUG_ALL,
	})
	serialized, err := Serialize(bs)
	if err != nil {
		t.Fatalf("Serialize failed: %v", err)
	}
	tests := []struct {
		name    string
		in      []byte
		t       time.Time
		wantErr error
	}{
		{
			name: "success",
			in:   serialized,
			t:    time.Now(),
		},
		{
			name:    "expiry",
			in:      serialized,
			t:       time.Unix(0, 0),
			wantErr: status.ErrInvalidArgument,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			gotErr := ValidateMetadataCardinality(tc.in, tc.t)
			if !errors.Is(gotErr, tc.wantErr) {
				t.Errorf("ValidateMetadataCardinality() returned error: %v, want error: %v", gotErr, tc.wantErr)
			}
		})
	}
}

func TestRoundTripV1(t *testing.T) {
	bs := New(&NewBinaryFields{
		Version:     1,
		Country:     "US",
		Region:      "US-CA",
		City:        "SUNNYVALE",
		ServiceType: "chromeipblinding",
		Expiration:  &tpb.Timestamp{Seconds: 3600},
		DebugMode:   pmpb.PublicMetadata_DEBUG_ALL,
	})
	serialized, err := Serialize(bs)
	if err != nil {
		t.Fatalf("Serialize failed: %v", err)
	}
	deserialized, err := Deserialize(serialized)
	if err != nil {
		t.Fatalf("Deserialize failed: %v", err)
	}
	if bs.GetServiceType() != deserialized.GetServiceType() {
		t.Errorf("service_type: got %v; want %v", deserialized.GetServiceType(), bs.GetServiceType())
	}
	if bs.GetExpiration().GetSeconds() != deserialized.GetExpiration().GetSeconds() {
		t.Errorf("expiration: got %v; want %v", deserialized.GetExpiration().GetSeconds(), bs.GetExpiration().GetSeconds())
	}
	if bs.GetDebugMode() != deserialized.GetDebugMode() {
		t.Errorf("debug_mode: got %v; want %v", deserialized.GetDebugMode(), bs.GetDebugMode())
	}
	if bs.GetExitLocation().GetCountry() != deserialized.GetExitLocation().GetCountry() {
		t.Errorf("country: got %v; want %v", deserialized.GetExitLocation().GetCountry(), bs.GetExitLocation().GetCountry())
	}
	if bs.metadata.GetRegion().Value() != deserialized.metadata.GetRegion().Value() {
		t.Errorf("region: got %q; want %q", deserialized.metadata.GetRegion().Value(), bs.metadata.GetRegion().Value())
	}
	if bs.metadata.GetCity().Value() != deserialized.metadata.GetCity().Value() {
		t.Errorf("city: got %q; want %q", deserialized.metadata.GetCity().Value(), bs.metadata.GetCity().Value())
	}
	if bs.GetProxyLayer() != deserialized.GetProxyLayer() {
		t.Errorf("proxy_layer: got %v; want %v", deserialized.GetProxyLayer(), bs.GetProxyLayer())
	}
}

func TestRoundTripV2(t *testing.T) {
	bs := New(&NewBinaryFields{
		Version:     2,
		Country:     "US",
		Region:      "US-CA",
		City:        "SUNNYVALE",
		ServiceType: "chromeipblinding",
		Expiration:  &tpb.Timestamp{Seconds: 3600},
		DebugMode:   pmpb.PublicMetadata_DEBUG_ALL,
		ProxyLayer:  plpb.ProxyLayer_PROXY_B,
	})
	serialized, err := Serialize(bs)
	if err != nil {
		t.Fatalf("Serialize failed: %v", err)
	}
	deserialized, err := Deserialize(serialized)
	if err != nil {
		t.Fatalf("Deserialize failed: %v", err)
	}
	if bs.GetServiceType() != deserialized.GetServiceType() {
		t.Errorf("service_type: got %v; want %v", deserialized.GetServiceType(), bs.GetServiceType())
	}
	if bs.GetExpiration().GetSeconds() != deserialized.GetExpiration().GetSeconds() {
		t.Errorf("expiration: got %v; want %v", deserialized.GetExpiration().GetSeconds(), bs.GetExpiration().GetSeconds())
	}
	if bs.GetDebugMode() != deserialized.GetDebugMode() {
		t.Errorf("debug_mode: got %v; want %v", deserialized.GetDebugMode(), bs.GetDebugMode())
	}
	if bs.GetExitLocation().GetCountry() != deserialized.GetExitLocation().GetCountry() {
		t.Errorf("country: got %v; want %v", deserialized.GetExitLocation().GetCountry(), bs.GetExitLocation().GetCountry())
	}
	if bs.metadata.GetRegion().Value() != deserialized.metadata.GetRegion().Value() {
		t.Errorf("region: got %q; want %q", deserialized.metadata.GetRegion().Value(), bs.metadata.GetRegion().Value())
	}
	if bs.metadata.GetCity().Value() != deserialized.metadata.GetCity().Value() {
		t.Errorf("city: got %q; want %q", deserialized.metadata.GetCity().Value(), bs.metadata.GetCity().Value())
	}
	if bs.GetProxyLayer() != deserialized.GetProxyLayer() {
		t.Errorf("proxy_layer: got %v; want %v", deserialized.GetProxyLayer(), bs.GetProxyLayer())
	}
}
